### PR TITLE
[MPS] Add workaround for nonzero with large/complex inputs 

### PR DESCRIPTION
**Pull request by skotapati**
Fixes Issue #122916

Resolves correctness issue seen with large inputs to the mps nonzero op by using a different scatter mode. Native nonzero op is still used with smaller inputs for better performance.

### CODE REVIEW

**Comment by kulinseth** [comment]
@skotapati can you add a test case ?

**Review by kulinseth — APPROVED** [review]

File: aten/src/ATen/native/mps/operations/Indexing.mm [hunk 1]
```diff
@@ -244,6 +246,57 @@ static Tensor nonzero_fallback(const Tensor& self) {
...
+
+  using namespace mps;
+
+  int64_t nDim = self.dim();
+  MPSStream* stream = getCurrentMPSStream();
+  struct CachedGraph : public MPSCachedGraph {
+    CachedGraph(MPSGraph* graph) : MPSCachedGraph(graph) {}
+    MPSGraphTensor* inputTensor_ = nil;
+    MPSGraphTensor* outputTensor_ = nil;
+  };
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/Indexing.mm:259]: This is a regular unary op, so let's use a predefined structure from OperationUtils.h
```suggestion
  using CachedGraph = MPSUnaryCachedGraph;
```

File: aten/src/ATen/native/mps/operations/Indexing.mm [hunk 2]
```diff
@@ -244,6 +246,57 @@ static Tensor nonzero_fallback(const Tensor& self) {
...
+  int64_t total_nonzero = at::count_nonzero(self).item<int64_t>();
+  at::native::resize_output(out_, {total_nonzero, nDim});
+  if (out_.numel() == 0) {
+    return out_;
+  }
+
+  bool contiguous_output = !needsGather(out_);
+  Tensor out = out_;
+  if (!contiguous_output) {
+    out = at::empty(out_.sizes(), out_.scalar_type(), c10::nullopt, kMPS, c10::nullopt, c10::nullopt);
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/Indexing.mm:273]:
```suggestion
    out = at::empty_like(out_, MemoryFormat::Contiguous);
```

File: aten/src/ATen/native/mps/operations/Indexing.mm [hunk 3]
```diff
@@ -244,6 +246,57 @@ static Tensor nonzero_fallback(const Tensor& self) {
...
+  }
+
+  @autoreleasepool {
+    string key = "nonzero_out_native_mps" + getTensorsStringKey(self);
+    auto cachedGraph = LookUpOrCreateCachedGraph<CachedGraph>(key, [&](auto mpsGraph, auto newCachedGraph) {
+      MPSGraphTensor* inputTensor = mpsGraphRankedPlaceHolder(mpsGraph, getMPSDataType(self), getMPSShape(self));
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/Indexing.mm:279]: Let the compiler do the work for you :)
```suggestion
      MPSGraphTensor* inputTensor = mpsGraphRankedPlaceHolder(mpsGraph, self);
```

File: aten/src/ATen/native/mps/operations/Indexing.mm [hunk 4]
```diff
@@ -305,20 +359,91 @@ static Tensor nonzero_fallback(const Tensor& self) {
     out = at::empty(out_.sizes(), out_.scalar_type(), c10::nullopt, kMPS, c10::nullopt, c10::nullopt);
   }
 
+  int64_t _apparentInputShape = 1;
+  for (auto dim : self.sizes()) {
+    _apparentInputShape *= dim;
+  }
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/Indexing.mm:365]: You are computing number of elements there, aren't you?
```suggestion
  int64_t _apparentInputShape = self.numel();
```

File: aten/src/ATen/native/mps/operations/Indexing.mm [hunk 5]
```diff
@@ -305,20 +359,91 @@ static Tensor nonzero_fallback(const Tensor& self) {
...
+      MPSGraphTensor* scatterDataTensor = mpsGraphUnrankedPlaceHolder(mpsGraph, getMPSScalarType(out.scalar_type()));
+      MPSGraphTensor* zeroTensor = [mpsGraph constantWithScalar:0.0 dataType:inputDataType];
+      MPSGraphTensor* oneTensor = [mpsGraph constantWithScalar:1.0 dataType:MPSDataTypeInt32];
+      MPSGraphTensor* minusMaxDimTensor = [mpsGraph constantWithScalar:-maxDimensions dataType:MPSDataTypeInt32];
+      MPSGraphTensor* inputNotEqualToZeroTensor = [mpsGraph notEqualWithPrimaryTensor:inputTensor
+                                                                      secondaryTensor:zeroTensor
+                                                                                 name:nil];
+      MPSGraphTensor* maskTensor = [mpsGraph castTensor:inputNotEqualToZeroTensor
+                                                 toType:MPSDataTypeInt32
+                                                   name:@"castToInt32"];
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/Indexing.mm:395]: Nit
```suggestion
      MPSGraphTensor* maskTensor = castMPSTensor(mpsGraph, inputNotEqualToZeroTensor, kInt);
```

**Review by malfet — APPROVED** [review]
LGTM, but please consider using more convenience functions

File: aten/src/ATen/native/mps/operations/Indexing.mm [hunk 6]
```diff
@@ -42,6 +42,8 @@
 #include <ATen/ops/view_as_real.h>
 #endif
 
+#define NONZERO_MAX_SIZE 16777216 // 2 ^ 24
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/Indexing.mm:45]: Not, define is so C89, why not `constexpr auto nonZeroMaxSize == 1UL << 24;

File: test/test_mps.py [hunk 7]
```diff
@@ -10586,6 +10586,7 @@ def helper(dtype):
                 torch.Size((6, 2)),
                 torch.Size((3, 2, 2)),
                 torch.Size((5, 5, 5)),
+                torch.Size((20000, 10000)),
```
Comment (malfet) [inline test/test_mps.py:10589]: How much this will increase test runtime? Perhaps it could be skipped?

**Pull request closed without merging**